### PR TITLE
add 'Runtime' struct to make test easier; add create test

### PR DESCRIPTION
**Pull request by liangchenye**
Signed-off-by: Liang Chenye <liangchenye@huawei.com>

### CODE REVIEW

**Comment by liangchenye** [comment]
Hi @Mashimiao , this is my approach to test lifecycle.
We need more test cases to cover 'create' operation. 
Mark as WIP now.

File: validation/validation_test.go [hunk 1]
```diff
@@ -11,17 +11,21 @@ import (
...
 
 var (
 	runtime = "runc"
 )
 
 func init() {
-	runtime = os.Getenv("RUNTIME")
+	runtimeInEnv := os.Getenv("RUNTIME")
+	if runtimeInEnv != "" {
+		runtime = runtimeInEnv
```
Comment (wking) [inline validation/validation_test.go]: This looks like a non-WIP improvement to me.  Can we move it into its own PR to get it landed more quickly?

File: validation/validation_test.go [hunk 2]
```diff
@@ -11,17 +11,21 @@ import (
...
+	}
 }
 
-func runtimeValidate(runtime string, g *generate.Generator) error {
+func runtimeValidate(runtime string, g *generate.Generator, args []string) error {
```
Comment (wking) [inline validation/validation_test.go]: This is not going to work, because when you test `start`, `state`, `delete` etc. you'll need to call `create` first (in most cases).  `runtimeValidate` contains all the bundle setup and teardown code, so you need a way to call multiple runtime commands inside the test, not just a way to change which single runtime command you call.  Pulling the bundle creation and teardown functionality out into separate functions seems like a more flexible approach, allowing callers to easily do whatever they want in between the two.

File: validation/validation_test.go (hunk 1 above)
Comment (liangchenye) [reply validation/validation_test.go]: https://github.com/opencontainers/runtime-tools/pull/448 is created.

File: validation/validation_test.go (hunk 2 above)
Comment (liangchenye) [reply validation/validation_test.go]: Your are right, move 'create/start/state/delete' to 'args' is not enough. 
https://github.com/opencontainers/runtime-tools/pull/391/  @Mashimiao  is the way to prepare the bundle.

**Comment by liangchenye** [comment]
It is similar with @Mashimiao 's https://github.com/opencontainers/runtime-tools/pull/391.
I just add a 'Runtime' struct and add more test cases in 'create' test.

In current testing, `runc` does not follow the lifecycle spec...

**Comment by liangchenye** [comment]
Updated, PTAL @wking @Mashimiao @mrunalp @q384566678

**Comment by liangchenye** [comment]
change to [WIP] . the `stretchr/testify` pkg is not included.
I'll add it in https://github.com/opencontainers/runtime-tools/pull/451

**Comment by liangchenye** [comment]
Updated with #451 rfc code feature.
PTAL @q384566678 @Mashimiao

**Comment by Mashimiao** [comment]
LGTM

**Comment by q384566678** [comment]
LGTM

File: validation/validation_test.go [hunk 3]
```diff
@@ -24,81 +27,106 @@ func init() {
...
-
-	return nil
-}
-
-func getDefaultGenerator() *generate.Generator {
-	g := generate.New()
-	g.SetRootPath(".")
-	g.SetProcessArgs([]string{"/runtimetest"})
-	return &g
+	return r.Start()
```
Comment (wking) [inline validation/validation_test.go]: This is broken, since you've deferred a `delete` call (via `Clean`) [earlier in the function][1].  That doesn't work reliably though, because [`delete` MUST generate an error if the container is not `stopped`][2], and the fact that `start` completes doesn't mean you're `stopped` (you [could also be `running`][3]).  In order for `Clean()` to work reliably, you either need to attempt to send a `KILL` to the container process with [`kill`][4] (although the spec is not clear on what signals are supported, you need something like #321 to see that [`KILL` MUST be supported][5]) or code that waits for the container process to exit on its own (more on this [here][6], in #305, and in most of the other PRs that attempted to split `run` into `create` and`start`, e.g. [here][7] and [here][8]).

[1]: https://github.com/opencontainers/runtime-tools/pull/447/files#diff-b5e8ff579203d652582cf33ed2b39ad9R65
[2]: https://github.com/opencontainers/runtime-spec/blob/v1.0.0/runtime.md#delete
[3]: https://github.com/opencontainers/runtime-spec/blame/v1.0.0/runtime.md#L21-L22
[4]: https://github.com/opencontainers/runtime-spec/blob/v1.0.0/runtime.md#kill
[5]: https://github.com/wking/ocitools-v2/blob/a2725b42d7b312681ed5315cf7e663a13c487445/docs/command-line-interface.md#kill
[6]: https://github.com/wking/ocitools-v2/blob/a2725b42d7b312681ed5315cf7e663a13c487445/docs/command-line-interface.md#container-process-exit
[7]: https://github.com/opencontainers/runtime-tools/pull/232#discussion_r82655833
[8]: https://github.com/opencontainers/runtime-tools/pull/285#issuecomment-264515242
